text/3649-macros-named-capture-groups.md

```diff
@@ -0,0 +1,799 @@
+- Feature Name: `macros-named-capture-groups`
+- Start Date: 2024-05-28
+- RFC PR: [rust-lang/rfcs#3649](https://github.com/rust-lang/rfcs/pull/3649)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+It will now be possible to give names to capture (repetition) groups in macro
+patterns, which can then be referred to directly in the macro body and macro
+metavariable expressions.
+
+_Rustc usually refers to these groups as "repetitions" in diagnostics. This
+RFC uses "capture groups" which is more general (they don't always repeat),
+and more in line with regex._
+
+# Motivation
+[motivation]: #motivation
+
+Rust has no way to refer to capture groups directly, so it uses the variables
+they capture to refer to them indirectly. This leads to confusing or limited
+behavior in a few places:
+
+- Expansion with multiple capture groups is extremely limited. In many cases,
+  the ordering and nesting of different groups is restricted based on what can
+  be inferred by the contained variables, since the groups themselves are
+  ambiguous.
+- Repetition-related diagnostics are suboptimal because the compiler has limited
+  ability to guess what a capture group _should_ refer to when a captured
+  groups and variables do not align correctly.
+- Repetition mismatch diagnostics can only be emitted after the macro is
+  instantiated, rather than when the macro is written. (E.g. "meta-variable
+  `foo` repeats 2 times, but `bar` repeats 1 time")
+- As a result of the above, using repetition is somewhat fragile; small
+  adjustments can break working patterns with little indication of what exactly
+  is wrong. Reading code with multiple capture groups can also be confusing.
+- Metavariable expressions as they currently exist use an unintuitive format:
+  syntax like `${count($var, n)}` is used to refer to the `n`th ancestor group
+  of the smallest group that captures `$var`. Referring to groups directly would
+  be more straightforward than using a proxy.
+
+It is expected that named capture groups will provide a way to remove ambiguity
+in expansion and metavariable expressions, as well as unblock diagnostics that
+do a better job of guiding the macro mental model.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Capture groups can now take a name by providing an identifier between the `$`
+and opening `(`. This group can then be referred to by name in the expansion:
+
+```rust
+macro_rules! foo {
+    ( $group1( $a:ident ),+ ) => {
+        $group1( println!("{}", $a); )+
+    }
+}
+```
```

Comment:
As mentioned by other reviewers, the expansion-side macro is already valid.

I'd like to suggest an alternative:

```rust
macro_rules! foo {
    ( $group1( $a:ident ),+ ) => {

        ${for a in group1} (
            println!("{}", $a);
        )
    }
}
```

This alternative is more verbose, but it's also much more explicit about what's going on.